HeavyFlavorAnalysis/SpecificDecay/plugins/BPHHistoSpecificDecay.cc

```diff
@@ -0,0 +1,740 @@
+    const reco::Candidate* kptr = BPHDaughters::get( cand, 0.49, 0.50 ).front();
+    if ( kptr == 0 ) continue;
+    if ( kptr->pt() < buKPtMin ) continue;
+    fillHisto( "Bu"    ,  cand );
+    fillHisto( "BuJPsi", *jPsi );
+  }
+
+  //////////// Bd ////////////
+
+  edm::Handle< vector<pat::CompositeCandidate> > bdCands;
+  bdCandsToken.get( ev, bdCands );
+
+  int ibd;
+  int nbd = bdCands->size();
+  for ( ibd = 0; ibd < nbd; ++ ibd ) {
+    LogTrace( "DataDump" )
+           << "*********** Bd " << ibd << "/" << nbd << " ***********";
+    const pat::CompositeCandidate& cand = bdCands->at( ibd );
+    const pat::CompositeCandidate* jPsi = BPHUserData::getByRef
+         <pat::CompositeCandidate>( cand, "refToJPsi" );
+    LogTrace( "DataDump" )
+           << "JPsi: " << jPsi;
+    if ( jPsi == 0 ) continue;
+    const pat::CompositeCandidate* kx0 = BPHUserData::getByRef
+         <pat::CompositeCandidate>( cand, "refToKx0" );
+    LogTrace( "DataDump" )
+           << "Kx0: " << kx0;
+    if ( kx0 == 0 ) continue;
+    if ( !bdJPsiBasicSelect   ->accept( *jPsi ) ) continue;
+    if ( !bdKx0BasicSelect    ->accept( * kx0 ) ) continue;
+    if ( !bdJPsiDaughterSelect->accept( *jPsi ) ) continue;
+    if ( !bdVertexSelect->accept( cand,
+                                  BPHUserData::getByRef<reco::Vertex>( *jPsi,
+                                  "primaryVertex" ) ) ) continue;
+    fillHisto( "Bd"    ,  cand );
+    fillHisto( "BdJPsi", *jPsi );
+    fillHisto( "BdKx0" , *kx0  );
+  }
+
+  //////////// Bs ////////////
+
+  edm::Handle< vector<pat::CompositeCandidate> > bsCands;
+  bsCandsToken.get( ev, bsCands );
+
+  int ibs;
+  int nbs = bsCands->size();
+  for ( ibs = 0; ibs < nbs; ++ ibs ) {
+    LogTrace( "DataDump" )
+           << "*********** Bs " << ibs << "/" << nbs << " ***********";
+    const pat::CompositeCandidate& cand = bsCands->at( ibs );
+    const pat::CompositeCandidate* jPsi = BPHUserData::getByRef
+         <pat::CompositeCandidate>( cand, "refToJPsi" );
+    LogTrace( "DataDump" )
+           << "JPsi: " << jPsi;
+    if ( jPsi == 0 ) continue;
+    const pat::CompositeCandidate* phi = BPHUserData::getByRef
+         <pat::CompositeCandidate>( cand, "refToPhi" );
+    LogTrace( "DataDump" )
+           << "Phi: " << phi;
+    if ( phi == 0 ) continue;
+    if ( !bsJPsiBasicSelect   ->accept( *jPsi ) ) continue;
+    if ( !bsPhiBasicSelect    ->accept( * phi ) ) continue;
+    if ( !bsJPsiDaughterSelect->accept( *jPsi ) ) continue;
+    if ( !bsVertexSelect->accept( cand,
+                                  BPHUserData::getByRef<reco::Vertex>( *jPsi,
+                                  "primaryVertex" ) ) ) continue;
+    fillHisto( "Bs"    ,  cand );
+    fillHisto( "BsJPsi", *jPsi );
+    fillHisto( "BsPhi" , *phi  );
+  }
+
+
+  return;
+
+}
+
+
+void BPHHistoSpecificDecay::endJob() {
+  TDirectory* currentDir = gDirectory;
+  TFile file( outHist.c_str(), "RECREATE" );
```

Comment:
code like this (direct root file and histogram management) is probably OK for analysis jobs running locally or with crab.
It can not be a part of a skim or some centrally running application